parquet-column/src/main/java/org/apache/parquet/column/statistics/SizeStatistics.java

```diff
@@ -0,0 +1,268 @@
+  private final List<Long> repetitionLevelHistogram;
+  /**
+   * Same as repetition_level_histogram except for definition levels.
+   *
+   * This field may be omitted if max_definition_level is 0 or 1 without
+   * loss of information.
+   */
+  private final List<Long> definitionLevelHistogram;
+
+  /**
+   * Whether the statistics has valid value.
+   *
+   * It is true by default. Only set to false while it fails to merge statistics.
+   */
+  private boolean hasValue = true;
+
+  /**
+   * Builder to create a SizeStatistics.
+   */
+  public static class Builder {
+    private final PrimitiveType type;
+    private long unencodedByteArrayDataBytes;
+    private final long[] repetitionLevelHistogram;
+    private final long[] definitionLevelHistogram;
+
+    /**
+     * Create a builder to create a SizeStatistics.
+     *
+     * @param type physical type of the column associated with this statistics
+     * @param maxRepetitionLevel maximum repetition level of the column
+     * @param maxDefinitionLevel maximum definition level of the column
+     */
+    public Builder(PrimitiveType type, int maxRepetitionLevel, int maxDefinitionLevel) {
+      this.type = type;
+      this.unencodedByteArrayDataBytes = 0L;
+      repetitionLevelHistogram = new long[maxRepetitionLevel + 1];
+      definitionLevelHistogram = new long[maxDefinitionLevel + 1];
+    }
+
+    /**
+     * Add repetition and definition level of a value to the statistics.
+     * It is called when value is null, or the column is not of BYTE_ARRAY type.
+     *
+     * @param repetitionLevel repetition level of the value
+     * @param definitionLevel definition level of the value
+     */
+    public void add(int repetitionLevel, int definitionLevel) {
+      Preconditions.checkArgument(
+          0 <= repetitionLevel && repetitionLevel < repetitionLevelHistogram.length,
+          "repetitionLevel %s is out of range [0, %s]",
+          repetitionLevel,
+          repetitionLevelHistogram.length - 1);
+      Preconditions.checkArgument(
+          definitionLevel < definitionLevelHistogram.length,
+          "definitionLevel %s is out of range [0, %s]",
+          definitionLevel,
+          definitionLevelHistogram.length - 1);
+      repetitionLevelHistogram[repetitionLevel]++;
+      definitionLevelHistogram[definitionLevel]++;
+    }
+
+    /**
+     * Add repetition and definition level of a value to the statistics.
+     * It is called when the column is of BYTE_ARRAY type.
+     *
+     * @param repetitionLevel repetition level of the value
+     * @param definitionLevel definition level of the value
+     * @param value value of to be added
+     */
+    public void add(int repetitionLevel, int definitionLevel, Binary value) {
+      add(repetitionLevel, definitionLevel);
+      if (type.getPrimitiveTypeName() == PrimitiveType.PrimitiveTypeName.BINARY && value != null) {
+        unencodedByteArrayDataBytes = Math.addExact(unencodedByteArrayDataBytes, value.length());
```

Comment:
I tend to keep this check just in case.

Reply:
In most cases, this good have checks like this one. But it can significantly hit performance when used in places called regularly. This method is invoked for every values. We shall be as effective as possible.

Reply:
Make sense. I have removed the overflow check.